17 決済機能新規作成

terukyou opened the pull request:
**正常な場合の挙動確認**
- [x] DBに挿入されるか確認(points、payments、members(total_point)
- [x] ポイントのセレクトボックスの挙動確認
- [x] 決済概要の計算が正しいか確認

**不正な場合の処理**
- [x] URL直打ち対策(セッションかあるかで判定
- [x] カードの値番号が不正の場合エラーに飛ぶ
- [x] すでにレコードにあるPK(列・行)が選択された場合はエラーに飛ぶ

**注意**
- [x] ポイントに関するバリデーション(model)を作成(利用可能ポイント>使用ポイントの判定はcontroller
- [x] ページリロード時に自動でポイントのselectのjs処理を実行
(「一部利用する」を選択した際は、入力した情報が次の画面に遷移した後もう一度ポイント入力画面に戻ると消えてしまう

hirayama-sachi committed:
add: 決済機能新規作成

hirayama-sachi committed:
remove: コメント化していない不要な文字が入っていたため削除した

checkpayment.ctpの不要な文字を削除

kazuyanakata commented on html/mycakeapp/src/Controller/ReservesController.php:
```diff
@@ -217,4 +228,163 @@ public function checkdetail()
+
+        $totalOfOwnPoints = $this->Members->find()
+            ->select(['total_point'])->where(['id' => $memberId])->first()->total_point;
+        $amountOfPayment = $_SESSION['fee'] - $_SESSION['displayed_amount'];
```
$_SESSION['displayed_amount']は引かれる額だけではなく、is_minusフラグが0の場合もあります。

①reservationsDetailsテーブルから当てはまるレコードを抜き出し、discount_idを抽出する
②discountsテーブルからdiscount_idのレコードを抜き出しis_minusカラムの値を確認する
③is_minusフラグが1の場合：$amountOfPayment = $_SESSION['fee'] - $_SESSION['displayed_amount'];
　is_minusフラグが0の場合：$amountOfPayment = $_SESSION['displayed_amount'];

の順で処理すると良いと思います！

terukyou replied:
確かにis_minusが立ってない場合の処理を考慮していないですね..
改善策も考えていただきありがとうございます!
修正いたします！！

terukyou replied:
5ca4bb1 こちらで修正いたしました

kazuyanakata commented on html/mycakeapp/src/Controller/ReservesController.php:
```diff
@@ -217,4 +228,163 @@ public function checkdetail()
+
+        $displayedAmount = $_SESSION['displayed_amount'];
+        $usePoint = $_SESSION['payment']['use_point'];
+        $subTotal = $fee - $displayedAmount - $usePoint;
```
上と同様の理由での修正依頼です。
discountのis_minusフラグが0の場合もあるのでその対応をお願いいたします。
対応方法は上記のレビューとほぼ同じかと思われます。

terukyou replied:
5ca4bb1 こちらで修正いたしました

kazuyanakata commented on html/mycakeapp/src/Controller/ReservesController.php:
```diff
@@ -217,4 +228,163 @@ public function checkdetail()
+        $displayedAmount = $_SESSION['displayed_amount'];
+        $usePoint = $_SESSION['payment']['use_point'];
+        $subTotal = $fee - $displayedAmount - $usePoint;
+        $taxRate = $this->Taxes->find()->select(['tax_rate'])->where(['id' => 1])->first()->tax_rate;
```
'id' => 1だと税率が変わった時にソースコードも変えなければいけないので、適用開始日時と適用終了日時で判断すべきだと思います！！そうするとDBテーブル変更だけで済むので！

①予約日が適用開始日時と適用終了日時の間
②予約日が適用開始日時過ぎててかつ適用終了日がnull

この二つの条件どちらかを満たす形にしたほうが良いと思います！

terukyou replied:
おっしゃる通りです..!!
こちらも修正いたします。

terukyou replied:
96dac5f こちらで修正いたしました

kazuyanakata commented on html/mycakeapp/src/Controller/ReservesController.php:
```diff
@@ -217,4 +228,163 @@ public function checkdetail()
+                'created_at' => $today,
+                'updated_at' => $today
+            ];
+            $pointRate = $this->PointRates->find()->select(['point_rate'])->where(['id' => 1])->first()->point_rate;
```
ポイント割合も税率と同様の理由です！
対応方法も同じかと思われます！

terukyou replied:
96dac5f こちらで修正いたしました

kazuyanakata commented on html/mycakeapp/src/Form/ReserveForm.php:
```diff
@@ -0,0 +1,30 @@
+<?php
```
<Thank you point!!!!>
モデルのないフォーム勉強になります！！ありがとうございます！

kazuyanakata commented on html/mycakeapp/webroot/css/payment.css:
```diff
@@ -0,0 +1,57 @@
+.error-message:not(:first-child),
+.error-message li:not(:first-child) {
+    padding-top: 5px;
+}
```
```suggestion
}
input[type='radio'] {
    -webkit-appearance: none;
    -moz-appearance: none;
    -ms-appearance: none;
    -o-appearance: none;
    appearance: none;
    position: relative;
    top: 25px;
    right: 0;
    bottom: 0;
    left: 0;
    height: 20px;
    width: 20px;
    vertical-align: -0.6rem;
    transition:all .15s ease-out 0s;
    color: #fff;
    cursor: pointer;
    display: inline-block;
    margin: .4rem;
    outline: none;
    border-radius: 10%;
  }
  input[type=radio] {
    border-radius: 50%;
  }
  input[type=radio]:checked:before {
    transform: scale(1);
  }
  input[type=radio]:before {
    content: "";
    display: block;
    width: 10px;
    height: 10px;
    border-radius: 50%;
    margin: 3px;
    transform: scale(0);
    transition: all ease-out 250ms;
  }
  input[type=radio] {
    border: 2px solid #F5AE0B;
  }
  input[type=radio]:checked:before {
    background: #F5AE0B;
  }

```

これ一番後ろに入れると、ラジオボタンがアニメーションの効いたいい感じのものになって、設計書の形に近づくのでよかったら是非！

terukyou replied:
ありがとうございます😭🙏
CSS苦手なので助かります🙇‍♂️🙇‍♂️

kazuyanakata commented on html/mycakeapp/webroot/css/payment.css:
```diff
@@ -0,0 +1,57 @@
+    margin: 32px 24px;
+}
+.cardInfo div {
+    padding: 20px 0 22px;
```
```suggestion
    padding: 20px 0 22px;
    cursor: pointer;
```
ここのカーソル変えるとユーザーもわかりやすくなると思います！

terukyou replied:
UI向上しますね!
ありがとうございます、修正します！！

kazuyanakata commented on html/mycakeapp/src/Template/Reserves/payment.ctp:
```diff
@@ -0,0 +1,43 @@
+</div>
+<div class="flex division-button">
+    <?php
+    echo $this->Html->link('キャンセル', ['action' => 'checkdetail'], ['class' => 'button normal-button back-gray']);
```
①adobe XDではキャンセルで基本設計書では戻るになってますね。ここ
予約詳細も全て戻るにしているので戻るで統一させましょう。
キャンセルだと予約自体全部キャンセルされると思われるかもなので！
②キャンセルボタンを押すとエラーになってしまいます。おそらくこの部分は次のフェーズでやると思うので問　題ないかと思うのですが一応。
③キャンセルボタンを押すことによって該当するreservationDetilsのレコードは削除しなきゃだと思うので、そこの対応はお願いいたします
　

terukyou replied:
こちらで修正いたしました
① af343c1
③ 69332b6
(②は次のフェーズで対応するので、③の確認の際はURL直打ち対策のコードをコメントアウトしてレビューお願します)

kazuyanakata commented on html/mycakeapp/src/Template/Reserves/checkpayment.ctp:
```diff
@@ -0,0 +1,37 @@
+</div>
+<div class="flex division-button">
+    <?php
+    echo $this->Html->link('キャンセル', ['action' => 'payment'], ['class' => 'button normal-button back-gray']);
```
ここも同様に戻るにしましょう

terukyou replied:
こちらで修正いたしました af343c1

kazuyanakata commented on html/mycakeapp/src/Controller/ReservesController.php:
```diff
@@ -217,4 +228,163 @@ public function checkdetail()
+                'updated_at' => $today
+            ];
+            $pointRate = $this->PointRates->find()->select(['point_rate'])->where(['id' => 1])->first()->point_rate;
+            $givePoint = ($fee * ($pointRate / 100));
```
付与ポイントは決済額の10％なので(基本設計書P18参照)
```suggestion
            $givePoint = floor(($purchasePrice * ($pointRate / 100)));
```
が正しい形だと思います！

terukyou replied:
こちらで修正いたしました 793f381

kazuyanakata commented on html/mycakeapp/src/Controller/ReservesController.php:
```diff
@@ -217,4 +228,163 @@ public function checkdetail()
+
+        if ($this->request->is('post')) {
+            // カード
+            $registeredOrder = $this->request->data['registered_order']; //カードを登録した順番(1番目、2番目)
+            if ($registeredOrder > 2) { //不正なカードの時(2枚以上登録できないはず)
+                $this->request->session()->delete('checkdetail');
+                $this->request->session()->delete('payment');
+                return $this->redirect(['controller' => 'error']);
+            }
```
以下の方法で送信するとエラー文が発生します
①開発者ツールでvalueを消した時
②カードが何も登録されていない時
③カードが一枚しか登録していない時に、開発者ツールでvalueを1または2に変更した時
④カードを2枚登録している時に、開発者ツールでvalueを2に変更した時
⑤valueに数字以外の値を入れた時

おそらく、if ($registeredOrder > 2)だけでは制限が緩いと思うので対応お願いいたします。

terukyou replied:
こちらで修正いたしました
①,⑤ 142e23c
② c6837f0
③,④ a35a769

kazuyanakata commented on html/mycakeapp/src/Controller/ReservesController.php:
```diff
@@ -217,4 +228,163 @@ public function checkdetail()
+        $_SESSION['column_number'] = 'A';
+        $_SESSION['record_number'] = 4;
+        $_SESSION['fee'] = 1800;
+        $_SESSION['displayed_amount'] = 300;
+        $_SESSION['discount_name'] = 'レディースデイ割引';
```
割引額と割引名をセッションで渡す前提となっていますが、割引の該当なしの場合はどのような形で渡すのでしょうか？
仮にセッションを渡さない方法だと、全ての画面で変数が定義されていない旨のエラー文が表示されてしまいます。
何か他の対応方法がもしあるのならお伺いできればと思います！！

terukyou replied:
こちらで修正いたしました 5ca4bb1

terukyou replied:
5ca4bb1
割引IDを引き継ぎ、クエリビルダーを用いて何も取得できなかった場合は割引なしとして対応する形へ変更しました！

kazuyanakata reviewed (commented):
レビュー完了しました！
コメントのご確認お願いいたします！

hirayama-sachi committed:
fix: 前の画面に戻るボタンを「戻る」へ修正

決済の方法と概要のボタンがXDと基本設計書で表示が違い、予約詳細が「戻る」に設定しており統一するため修正した

hirayama-sachi committed:
fix: 付与ポイントを決済額から10%へ修正

基本設計書(p.18参照 に決済額から10%を付与すると記載しているので修正した

hirayama-sachi committed:
fix: 割引の引継ぎをセッションにdiscounts.idを入れることで対応するよう修正

割引の該当なしやis_minusフラグが立っていない(入場料)場合に対応できない引継ぎ方法のため、割引IDで引継ぎ対応するよう修正した

hirayama-sachi committed:
fix: 消費税率とポイント付与率を適用開始日時と適用終了日時で判定するよう修正

値の変更があった際はDB修正のみで済むようにするため

hirayama-sachi committed:
add: カードのバリデーションを追加した(カードの値を意図的に空欄、文字列にした場合に対処するため

カードのバリデーションを追加し、バリデーションをpost送信直後に判定するように修正した

hirayama-sachi committed:
fix: カードの所有数と比較して不正なカードが判定するよう修正

一枚のみ登録している時、2枚目のカードを選ぶとバグとなってしまうためカード所有数によって判定するよう修正した

hirayama-sachi committed:
add:決済方法選択時、カード未登録の場合は新規登録ページへ遷移し登録後は決済方法へ遷移する機能追加

・カード未登録時に決済方法から決済概要へ進もうとするとエラーが出るバグが発生していた
・基本設計書(p.58に未登録の場合は新規登録で登録を経て決済方法を選択するようになっていた
上記2項目の理由より機能を追加した

hirayama-sachi committed:
add: 決済方法から前の画面に戻る際、予約詳細のキャンセルフラグをたて、予約詳細情報をDBから取得

terukyou committed:
Update html/mycakeapp/webroot/css/payment.css

Co-authored-by: kazuyanakata <68310528+kazuyanakata@users.noreply.github.com>

terukyou committed:
UI向上のためカード選択時はカーソルをリンクカーソルにする

Co-authored-by: kazuyanakata <68310528+kazuyanakata@users.noreply.github.com>

kazuyanakata commented on html/mycakeapp/src/Controller/ReservesController.php:
```diff
@@ -217,4 +272,193 @@ public function checkdetail()
+        $_SESSION['checkdetail'] = 1;
+        if (empty($_SESSION['checkdetail'])) {
+            return $this->redirect(['controller' => 'error']);
+            $this->request->session()->delete('checkdetail');
```
$_SESSION['checkdetail']がない場合の処理なので、ここのセッション削除は不要かと思います！

terukyou replied:
94ac69e
確かに...!!
ありがとうございます、修正いたしました!

kazuyanakata commented on html/mycakeapp/src/Controller/ReservesController.php:
```diff
@@ -161,6 +175,47 @@ public function discount()
     public function checkdetail()
     {
+        $memberId = $this->Auth->user('id');
+        $mainKey = [
+            'member_id' => $memberId,
+            'schedule_id' => $_SESSION['schedule_id'],
+            'column_number' => $_SESSION['column_number'],
+            'record_number' => $_SESSION['record_number'],
+        ];
+        $reservationDetails = $this->ReservationDetails->find('ReservationDetails', ['mainKey' => $mainKey]);
```
if文内で同じ処理をしている、かつこの変数はif文ないでしか使ってないので、ここは不要かと思います！

terukyou replied:
94ac69e
こちらも失礼しました!　修正いたしました。

kazuyanakata reviewed (commented):
お疲れ様です！
2点簡単な修正コメントを入れました！
これ修正したらLGTMかと！

hirayama-sachi committed:
Merge branch 'develop' of https://github.com/labotinc/codegym-team5 into feature/hinappu-17/#46

kazuyanakata commented on html/mycakeapp/src/Controller/ReservesController.php:
```diff
@@ -161,6 +175,47 @@ public function discount()
+            $_SESSION['fee'] = $this->Fees->get($_SESSION['detail']['fee_id']);
+        }
         //直接画面遷移の対応(途中でページを遷移したことによりセッションは残っていた場合も直接遷移させない)
         if (!(isset($_SESSION['detail']['discount_id'])) || $this->referer(null, true) !== '/reserves/discount' && $this->referer(null, true) !== '/reserves/checkdetail') {
             $this->request->session()->delete('seat');
```
一点修正追加です。今のままだと決済画面から戻るを押すとエラーが出るので
```suggestion
        if (!(isset($_SESSION['detail']['discount_id'])) || $this->referer(null, true) !== '/reserves/discount' && $this->referer(null, true) !== '/reserves/checkdetail' && $this->referer(null, true) !== '/reserves/payment') {
            $this->request->session()->delete('seat');
```
このように変更お願いします！

terukyou replied:
752e4db
いただいた指摘を取り入れたことに加え、条件文が非常に長くなったので、不正遷移の場合の変数を格納し、if文で値を調べる形へ変更しました。

hirayama-sachi committed:
remove:不要なコードの削除(決済方法と予約確認)

同じ処理が2度書かれている、空でないと入れないifで空のものを削除する等の無駄があったため削除した

hirayama-sachi committed:
add: 予約詳細に不正ページ遷移か調べる変数を格納、変数に決済方法を遷移可能に追加

決済方法から予約詳細に戻る際、エラーがでるバグがあったため遷移可能項目に追加した

terukyou commented:
> $_SESSION['checkdetail']がない場合の処理なので、ここのセッション削除は不要かと思います！ 

確かに...!!
ありがとうございます、修正いたしました! 94ac69e

kazuyanakata reviewed (commented):
LGTMです！！
マージお願いします！
